File: frontend/src/lib/validators.ts
// Email validation
export function validateEmail(email: string): { isValid: boolean; message?: string } {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/

  if (!email) {
    return { isValid: false, message: 'Email é obrigatório' }
  }

  if (!emailRegex.test(email)) {
    return { isValid: false, message: 'Formato de email inválido' }
  }

  // Check for disposable emails
  const disposableDomains = [
    '10minutemail.com',
    'tempmail.org',
    'guerrillamail.com',
    'mailinator.com',
    'throwaway.email',
    'temp-mail.org',
  ]

  const domain = email.split('@')[1].toLowerCase()
  if (disposableDomains.includes(domain)) {
    return { isValid: false, message: 'Emails temporários não são permitidos' }
  }

  return { isValid: true }
}

// Password validation
export function validatePassword(password: string): {
  isValid: boolean
  message?: string
  score: number
  suggestions: string[]
} {
  const suggestions: string[] = []
  let score = 0

  if (!password) {
    return { isValid: false, message: 'Senha é obrigatória', score: 0, suggestions: [] }
  }

  if (password.length < 8) {
    return {
      isValid: false,
      message: 'Senha deve ter pelo menos 8 caracteres',
      score: 0,
      suggestions: ['Adicione mais caracteres'],
    }
  }

  // Length check
  if (password.length >= 12) score += 2
  else if (password.length >= 8) score += 1

  // Character variety checks
  const hasUpper = /[A-Z]/.test(password)
  const hasLower = /[a-z]/.test(password)
  const hasDigit = /\d/.test(password)
  const hasSpecial = /[!@#$%^&*()_+\-=\[\]{}|;:,.<>?]/.test(password)

  if (!hasUpper) {
    suggestions.push('Adicione letras maiúsculas (A-Z)')
  } else {
    score += 1
  }

  if (!hasLower) {
    suggestions.push('Adicione letras minúsculas (a-z)')
  } else {
    score += 1
  }

  if (!hasDigit) {
    suggestions.push('Adicione números (0-9)')
  } else {
    score += 1
  }

  if (!hasSpecial) {
    suggestions.push('Adicione caracteres especiais (!@#$%^&*)')
  } else {
    score += 1
  }

  // Common password check
  const commonPasswords = [
    'password',
    '123456',
    '123456789',
    'qwerty',
    'abc123',
    'password123',
    'admin',
    'letmein',
    'welcome',
    'monkey',
  ]

  if (commonPasswords.includes(password.toLowerCase())) {
    return {
      isValid: false,
      message: 'Senha muito comum',
      score: 0,
      suggestions: ['Escolha uma senha mais única'],
    }
  }

  // Sequential characters check
  if (hasSequentialChars(password)) {
    score -= 1
    suggestions.push('Evite caracteres sequenciais (abc, 123)')
  }

  const isValid = score >= 3 && suggestions.length === 0

  return {
    isValid,
    message: isValid ? undefined : 'Senha não atende aos critérios de segurança',
    score,
    suggestions,
  }
}

// Phone validation
export function validatePhone(phone: string): {
  isValid: boolean
  message?: string
  formatted?: string
} {
  if (!phone) {
    return { isValid: true } // Phone is optional
  }

  // Remove all non-digit characters
  const digitsOnly = phone.replace(/\D/g, '')

  // Check if it's a valid length (7-15 digits)
  if (digitsOnly.length < 7 || digitsOnly.length > 15) {
    return { isValid: false, message: 'Número de telefone inválido' }
  }

  // Format phone number
  let formatted = phone
  if (digitsOnly.length === 10) {
    formatted = `(${digitsOnly.slice(0, 3)}) ${digitsOnly.slice(3, 6)}-${digitsOnly.slice(6)}`
  } else if (digitsOnly.length === 11 && digitsOnly[0] === '1') {
    formatted = `+1 (${digitsOnly.slice(1, 4)}) ${digitsOnly.slice(4, 7)}-${digitsOnly.slice(7)}`
  }

  return { isValid: true, formatted }
}

// Name validation
export function validateName(name: string): { isValid: boolean; message?: string } {
  if (!name || !name.trim()) {
    return { isValid: false, message: 'Nome é obrigatório' }
  }

  if (name.trim().length < 2) {
    return { isValid: false, message: 'Nome deve ter pelo menos 2 caracteres' }
  }

  if (name.trim().length > 200) {
    return { isValid: false, message: 'Nome deve ter no máximo 200 caracteres' }
  }

  // Check for valid characters (letters, spaces, hyphens, apostrophes)
  const nameRegex = /^[a-zA-ZÀ-ÿ\s\-']+$/
  if (!nameRegex.test(name.trim())) {
    return { isValid: false, message: 'Nome contém caracteres inválidos' }
  }

  return { isValid: true }
}

// Helper function to check for sequential characters
function hasSequentialChars(str: string): boolean {
  for (let i = 0; i < str.length - 2; i++) {
    const char1 = str.charCodeAt(i)
    const char2 = str.charCodeAt(i + 1)
    const char3 = str.charCodeAt(i + 2)

    if (char2 === char1 + 1 && char3 === char1 + 2) {
      return true
    }
  }
  return false
}

// Form validation helper
export function validateForm<T extends Record<string, any>>(
  data: T,
  rules: Record<keyof T, (value: any) => { isValid: boolean; message?: string }>,
): { isValid: boolean; errors: Record<keyof T, string> } {
  const errors: Record<keyof T, string> = {} as Record<keyof T, string>
  let isValid = true

  for (const [field, value] of Object.entries(data)) {
    const rule = rules[field as keyof T]
    if (rule) {
      const result = rule(value)
      if (!result.isValid) {
        errors[field as keyof T] = result.message || 'Campo inválido'
        isValid = false
      }
    }
  }

  return { isValid, errors }
}

File: backend/app/core/validators.py
import re
from typing import Any, Dict
from pydantic import BaseModel, validator
from email_validator import validate_email, EmailNotValidError


class EmailValidator:
    """Email validation utility."""
    
    @staticmethod
    def validate_email_format(email: str) -> bool:
        """Validate email format."""
        try:
            validate_email(email)
            return True
        except EmailNotValidError:
            return False
    
    @staticmethod
    def is_disposable_email(email: str) -> bool:
        """Check if email is from a disposable email service."""
        disposable_domains = {
            '10minutemail.com', 'tempmail.org', 'guerrillamail.com',
            'mailinator.com', 'throwaway.email', 'temp-mail.org'
        }
        domain = email.split('@')[1].lower()
        return domain in disposable_domains


class PasswordValidator:
    """Password validation utility."""
    
    @staticmethod
    def validate_strength(password: str) -> Dict[str, Any]:
        """Validate password strength and return detailed feedback."""
        feedback = {
            'is_valid': True,
            'score': 0,
            'issues': [],
            'suggestions': []
        }
        
        # Length check
        if len(password) < 8:
            feedback['is_valid'] = False
            feedback['issues'].append('Password must be at least 8 characters long')
        elif len(password) >= 12:
            feedback['score'] += 2
        
        # Character variety checks
        has_upper = any(c.isupper() for c in password)
        has_lower = any(c.islower() for c in password)
        has_digit = any(c.isdigit() for c in password)
        has_special = any(c in "!@#$%^&*()_+-=[]{}|;:,.<>?" for c in password)
        
        if not has_upper:
            feedback['issues'].append('Password must contain at least one uppercase letter')
            feedback['suggestions'].append('Add uppercase letters (A-Z)')
        else:
            feedback['score'] += 1
            
        if not has_lower:
            feedback['issues'].append('Password must contain at least one lowercase letter')
            feedback['suggestions'].append('Add lowercase letters (a-z)')
        else:
            feedback['score'] += 1
            
        if not has_digit:
            feedback['issues'].append('Password must contain at least one number')
            feedback['suggestions'].append('Add numbers (0-9)')
        else:
            feedback['score'] += 1
            
        if not has_special:
            feedback['issues'].append('Password must contain at least one special character')
            feedback['suggestions'].append('Add special characters (!@#$%^&*)')
        else:
            feedback['score'] += 1
        
        # Common password check
        common_passwords = {
            'password', '123456', '123456789', 'qwerty', 'abc123',
            'password123', 'admin', 'letmein', 'welcome', 'monkey'
        }
        
        if password.lower() in common_passwords:
            feedback['is_valid'] = False
            feedback['issues'].append('Password is too common')
            feedback['suggestions'].append('Choose a more unique password')
        
        # Sequential characters check
        if PasswordValidator._has_sequential_chars(password):
            feedback['score'] -= 1
            feedback['suggestions'].append('Avoid sequential characters (abc, 123)')
        
        return feedback
    
    @staticmethod
    def _has_sequential_chars(password: str) -> bool:
        """Check for sequential characters."""
        for i in range(len(password) - 2):
            if (ord(password[i+1]) == ord(password[i]) + 1 and 
                ord(password[i+2]) == ord(password[i]) + 2):
                return True
        return False


class PhoneValidator:
    """Phone number validation utility."""
    
    @staticmethod
    def validate_phone(phone: str) -> bool:
        """Validate phone number format."""
        # Remove all non-digit characters
        digits_only = re.sub(r'\D', '', phone)
        
        # Check if it's a valid length (7-15 digits)
        if len(digits_only) < 7 or len(digits_only) > 15:
            return False
        
        # Check for valid patterns
        patterns = [
            r'^\d{10}$',  # US format
            r'^\d{11}$',  # US with country code
            r'^\d{13}$',  # International
        ]
        
        return any(re.match(pattern, digits_only) for pattern in patterns)
    
    @staticmethod
    def format_phone(phone: str) -> str:
        """Format phone number to standard format."""
        digits_only = re.sub(r'\D', '', phone)
        
        if len(digits_only) == 10:
            return f"({digits_only[:3]}) {digits_only[3:6]}-{digits_only[6:]}"
        elif len(digits_only) == 11 and digits_only[0] == '1':
            return f"+1 ({digits_only[1:4]}) {digits_only[4:7]}-{digits_only[7:]}"
        
        return phone


class InputSanitizer:
    """Input sanitization utility."""
    
    @staticmethod
    def sanitize_string(value: str) -> str:
        """Sanitize string input."""
        if not isinstance(value, str):
            return str(value)
        
        # Remove null bytes and control characters
        sanitized = re.sub(r'[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]', '', value)
        
        # Trim whitespace
        sanitized = sanitized.strip()
        
        # Limit length
        return sanitized[:1000]
    
    @staticmethod
    def sanitize_email(email: str) -> str:
        """Sanitize email input."""
        if not isinstance(email, str):
            return ""
        
        # Convert to lowercase and trim
        sanitized = email.lower().strip()
        
        # Remove any whitespace
        sanitized = sanitized.replace(' ', '')
        
        return sanitized
